worker/uniter/resolver/loop.go

```diff
@@ -259,8 +259,12 @@ func checkCharmUpgrade(logger Logger, charmDir string, remote remotestate.Snapsh
 	local := rf.LocalState
 	local.State = ex.State()
 
-	// If the unit isn't installed, no need to start an upgrade.
-	if !local.State.Installed || remote.CharmURL == "" {
+	opFunc := rf.NewUpgrade
+	if !local.Installed && local.Kind == operation.RunHook && local.Step == operation.Pending && local.Hook.Kind == hooks.Install {
```

Comment:
This can just be
`if !local.Installed && local.Hook != nil && local.Hook.Kind == hooks.Install && local.Step != operation.Done {`

A not yet run hook and can pending or queued.
Hook !- nil is sufficient to check that an incomplete hook is due to be run.